src/util/strings.rs

```diff
@@ -91,84 +102,97 @@ impl StacksMessageCodec for StacksString {
-    }
-    if (buf.len() as u32) < index + len {
-        return Err(net_error::UnderflowError(format!("Not enough bytes to read string (have {}, need {})", buf.len(), index + len)));
-    }
-
-    let bytes : Vec<u8> = buf[(index as usize)..((index + len) as usize)].to_vec();
-    index += len;
-
-    *index_ptr = index;
-    Ok(bytes)
-}
-
 impl StacksMessageCodec for ClarityName {
-    fn serialize(&self) -> Vec<u8> {
-        let mut res = vec![];
+    fn consensus_serialize<W: Write>(&self, fd: &mut W) -> Result<(), net_error> {
         // ClarityName can't be longer than vm::representations::MAX_STRING_LEN, which itself is
         // a u8, so we should be good here.
-        assert!(self.as_bytes().len() <= CLARITY_MAX_STRING_LENGTH as usize);
-        res.push(self.as_bytes().len() as u8);
-        res.extend_from_slice(self.as_bytes());
-        res
+        if self.as_bytes().len() > CLARITY_MAX_STRING_LENGTH as usize {
+            return Err(net_error::SerializeError("Failed to serialize clarity name: too long".to_string()));
+        }
+        write_next(fd, &(self.as_bytes().len() as u8))?;
+        fd.write_all(self.as_bytes()).map_err(net_error::WriteError)?;
+        Ok(())
     }
 
-    fn deserialize(buf: &Vec<u8>, index_ptr: &mut u32, max_size: u32) -> Result<ClarityName, net_error> {
-        let mut index = *index_ptr;
-        let bytes = read_clarity_string_bytes(buf, &mut index, max_size)?;
+    fn consensus_deserialize<R: Read>(fd: &mut R) -> Result<ClarityName, net_error> {
+        let len_byte : u8 = read_next(fd)?;
+        if len_byte > CLARITY_MAX_STRING_LENGTH {
+            return Err(net_error::DeserializeError("Failed to deserialize clarity name: too long".to_string()));
+        }
+        let mut bytes = vec![0u8; len_byte as usize];
+        fd.read_exact(&mut bytes).map_err(net_error::ReadError)?;
 
         // must encode a valid string
         let s = String::from_utf8(bytes)
             .map_err(|_e| net_error::DeserializeError("Failed to parse Clarity name: could not contruct from utf8".to_string()))?;
 
         // must decode to a clarity name
         let name = ClarityName::try_from(s).map_err(|e| net_error::DeserializeError(format!("Failed to parse Clarity name: {:?}", e)))?;
-        *index_ptr = index;
         Ok(name)
     }
 }
 
 impl StacksMessageCodec for ContractName {
-    fn serialize(&self) -> Vec<u8> {
-        let mut res = vec![];
-        // ContractName can't be longer than vm::representations::MAX_STRING_LEN, which itself is
-        // a u8, so we should be good here.
-        assert!(self.as_bytes().len() <= CLARITY_MAX_STRING_LENGTH as usize);
-        res.push(self.as_bytes().len() as u8);
-        res.extend_from_slice(self.as_bytes());
-        res
+    fn consensus_serialize<W: Write>(&self, fd: &mut W) -> Result<(), net_error> {
+        if self.as_bytes().len() > CLARITY_MAX_STRING_LENGTH as usize {
```

Comment:
Should we introduce a constant CONTRACT_MAX_STRING_LENGTH (which is 40, not 128)?

Reply:
Ah, didn't know that.  Will fix.

Reply:
Should be fixed now.